feat(minifier): annotate more pure constructors

Boshen opened the pull request:
closes https://github.com/rolldown/rolldown/issues/2603

I used the following script to capture what argument throws.

```js
function x(F) {
    try {
    new F(null)
    } catch (e) {
        console.log(F.name)
    }
}
x(Set)
x(Map)
x(WeakSet)
x(WeakMap)
x(AggregateError)
x(Boolean)
x(DataView)
x(Error)
x(EvalError)
x(RangeError)
x(ReferenceError)
x(RegExp)
x(SyntaxError)
x(TypeError)
x(URIError)
x(ArrayBuffer)
x(Date)
x(Number)
x(Object)
x(String)
```

Boshen commented:
* **#11555** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11555?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11555?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11555).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented:
Some cases you may need to consider:
```js
new ArrayBuffer(-1)
new RegExp("(")
```

Boshen committed:
feat(minifier): annotate more pure constructors (#11555)

closes https://github.com/rolldown/rolldown/issues/2603

I used the following script to capture what argument throws.

```js
function x(F) {
    try {
    new F(null)
    } catch (e) {
        console.log(F.name)
    }
}
x(Set)
x(Map)
x(WeakSet)
x(WeakMap)
x(AggregateError)
x(Boolean)
x(DataView)
x(Error)
x(EvalError)
x(RangeError)
x(ReferenceError)
x(RegExp)
x(SyntaxError)
x(TypeError)
x(URIError)
x(ArrayBuffer)
x(Date)
x(Number)
x(Object)
x(String)
```
